### PR TITLE
Fix wrong error upper bound when performing incremental reductions

### LINKED ISSUE #40005: Error upper bound may be wrong when performing incremental reductions
When reducing terms aggs results, we check if we already have a doc count error for a certain bucket by looking at its error see https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java#L245): if it's greater than zero we have already calculated it, while if we have zero it means we have not hence we ignore such value and use the doc count of the last returned bucket.

When performing incremental reductions though, `0` may mean that the error was not previously calculated, or that the error was indeed previously calculated and its value was `0`. We end up rejecting true values set to `0` this way.

### PR DESCRIPTION
**Pull request by Hohol**
When performing incremental reductions, `0` value of `docCountError` may mean that the error was not previously calculated, or that the error was indeed previously calculated and its value was 0. We end up rejecting true values set to 0 this way.
This may lead to wrong upper bound of error in result.
To fix it, this PR makes `docCountError` nullable. `null` values mean that error was not calculated yet.

Fixes https://github.com/elastic/elasticsearch/issues/40005

### CODE REVIEW

**Comment by Hohol** [comment]
@jpountz, @jimczi, can I get some reviewers, please?

**Comment by polyfractal** [comment]
Apologies for the delay on a review, lots of folks have been in/out due to summer holidays :)

I'll take a look at it early this week, and I've also requested a review from @javanna when he gets back from holiday (also early this week).  Thanks for the PR @Hohol!

**Comment by Hohol** [comment]
I've updated the PR. Can we finish this?
Is the issue still relevant at all?

**Comment by Hohol** [comment]
Oops, looks like diff is unreadable now.

**Comment by Hohol** [comment]
Rebased.

**Comment by Hohol** [comment]
@polyfractal, @javanna, @colings86, can someone answer, please?

**Comment by imotov** [comment]
@elasticmachine test this please

**Comment by imotov** [comment]
@elasticmachine test this please

**Comment by imotov** [comment]
@elasticmachine test this please

**Comment by imotov** [comment]
@elasticmachine test this please

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedTerms.java [hunk 1]
```diff
@@ -51,7 +52,14 @@ protected InternalMappedTerms(String name, BucketOrder reduceOrder, BucketOrder
      */
     protected InternalMappedTerms(StreamInput in, Bucket.Reader<B> bucketReader) throws IOException {
         super(in);
-        docCountError = in.readZLong();
+        if (in.getVersion().onOrAfter(Version.V_8_0_0)) { // todo fix after backport
+            docCountError = in.readOptionalLong();
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedTerms.java:56]: Should this stay a zlong?

**Review by nik9000 — COMMENTED** [review]
I  *think* this is right. I'm really sorry we lost track of this one. It seems like a good change. I left a small question, but I think its right.

@imotov thought I knew this code fairly well. I'm not 100% confident in my ability to reason about it all in my head but with the test change and his confidence I'm happy with the change.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedTerms.java (hunk 1 above)
Comment (imotov) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedTerms.java:56]: @nik9000 why was it a zlong in the first place? Can errors be negative?

**Comment by nik9000** [comment]
Good question! Maybe it should be vlong? I'm not sure how it got that way
in the first place.

**Comment by imotov** [comment]
@elasticmachine update branch

**Comment by imotov** [comment]
@elasticmachine test this please

**Review by not-napoleon — APPROVED** [review]
I've read this twice now, and I think it's okay.  Very tricky.

**Comment by imotov** [comment]
@elasticmachine update branch

**Comment by imotov** [comment]
@elasticmachine test this please

**Comment by imotov** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by javanna** [comment]
I was expecting this to address an issue I found in CCSDuelIT, but that does not seem to be the case. See #85538 . We still have an awaitsfix and a failure when we try and remove it. Could you help with this?

**Pull request merged**